--- a/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2020-03-21 23:18:35.231267480 +0800
+++ b/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts	2020-03-21 23:21:01.266767176 +0800
@@ -197,9 +197,9 @@
 		compatible = "mediatek,eth-mac";
 		reg = <0>;
 		phy-mode = "trgmii";
-
+		local-mac-address = [00 0a 35 00 00 01];
 		fixed-link {
-			speed = <1000>;
+			speed = <2500>;
 			full-duplex;
 			pause;
 		};
@@ -211,6 +211,7 @@
 		label = "wan";
 		phy-mode = "rgmii";
 		phy-handle = <&ephy0>;
+		local-mac-address = [00 0a 35 00 00 02];
 	};
 
 	mdio: mdio-bus {
@@ -266,7 +267,7 @@
 					phy-mode = "trgmii";
 
 					fixed-link {
-						speed = <1000>;
+						speed = <2500>;
 						full-duplex;
 						pause;
 					};

